### ISSUE TITLE
Make 'return' keyword optional for last statement in function

### ISSUE THREAD

**Issue by masak** [masak/007#233]
Proposing this as a language feature while formally noting my reluctance and displeasure at the proposal. :stuck_out_tongue: 

Currently, [the homepage](http://masak.github.io/007/) boasts that 007, like Python but unlike Perl 6, eschews implicit returns. Maybe we should allow that, though.

Why? These are my three reasons:

* Currently there are no lambdas, and barring #215 the shortest way to write the identity function is `sub (v) { return v }`. Kind of long for such a natural idea. Would be nice if `sub (v) { v }` Just Worked.
* I keep *forgetting* to put in `return`, as if my brain expects the return to be implicit. (This also happens to me occasionally in JavaScript.)
* The hard-won conclusion of #7 means we *already* have this kind of semantics for quasis... so why not for functions too? Making it more (strangely) consistent.

**Comment by vendethiel** [comment]
Traverse sub-blocks (if/for)? Or do those not return anything except using a perl6-like do?

**Comment by masak** [comment]
The rule we ended up using for quasis (which I rather like) is that `Q::Statement::Expression` returns a value, but nothing else.

I think a `do` macro is near-possible in 007 nowadays. It should be implementable by what amounts almost to an identity macro:

    macro do(stmt) is parsed(/<statement>/) {
        return quasi @ Q::Statement {
            {{{Q::Statement @ stmt}}}
        }
    }

The idea being that wrapping `stmt` in a quasi like that makes it pick up the implicit return semantics of quasis.

**Comment by vendethiel** [comment]
That's... a bit scary, tbh

**Comment by masak** [comment]
But consistent.

**Comment by masak** [comment]
Coming back to the issue at hand, though &mdash; does it make sense to make `return` implicit in 007? And to give it the same semantics as with quasis, that is, hand back the value if it's a `Q::Statement::Expr`.

**Comment by masak** [comment]
The flip side of this strange consistency is that people might be *more* tempted to throw in `return` in a quasi:

    macro moo() {
        return quasi {
            return 42;
        };
    }

Their reasoning likely being, since the `return` on the macro/routine level is optional, the quasi will also work with or without a `return`.

But it won't work; `return` is a thing that binds to a surrounding routine, which in this case is the macro `moo`, which has already depleted once the `return` runs. It's perhaps possible for a `return` inside a quasi to work if that quasi is *not* part of a macro mechanism... but cases like the above will always be wrong, and perhaps deserve a compile-time error.

**Comment by masak** [comment]
*Or*...

...maybe since `return` is so useless in a quasi, one could pass an edict saying that `return` binds not just to routines, but to quasis, too. That is within our power as language designers, after all.

The advantage seems to be twofold. It would eliminate a sharp corner for people who expected `return` in a quasi to do something sensible; and it would perhaps even be a nice use case for some quasis to return early.

The disadvantage is that the rule for what `return` binds to now becomes more complex. But maybe it's easier than I thought to rationalize quasi being a routine of sorts... After all, it's a block with its own lexical pad.

**Comment by masak** [comment]
Been mulling over this one. It feels especially like [this comment](https://github.com/masak/007/issues/212#issuecomment-306496566) adds some kind of nail to some kind of coffin. I'm wondering if what we're looking at here isn't some kind of complete other consistent world view, yearning to break free from the current local optimum we're in.

To summarize, here's where we are right now:

1. Returning from routines need an explicit `return` keyword. Just dropping an expression at the last statement won't cut it. (What this issue is about.)
2. On the other hand, (the bits of code generated from) quasis have an implicit "return" semantics: whatever you put last in the quasi gets passed back to its surroundings.
3. The bits of code generated from quasis are not proper blocks or routines; they're `Q::Expr::StatementListAdapter`s.

Here's where we seem to be pushed:

1. Implicit returns from routines.
2. Implicit returns from (code generated from) quasis.
3. The code generated from quasis are *routines*, in a way.

Re that last point, this IIFEs of JavaScript fame:

    (function() {
        let x = "OH HAI";
        console.log(x);
    })();

Let's say we end up with a `Q::Expr::BlockAdapter`. That sounds reasonable. We don't need to include the actual `()` call in the Qtree, since that's implied. So a `Q::Expr::BlockAdapter` simply contains a `Q::Block`. That block will always have 0 parameters; unless we find some deviously clever (but safe!) way to also use those parameters to hygienically pass stuff around.

Anyway, tl;dr: the feedback from #212 on this makes me more convinced we should go ahead and make `return` implicit in 007, while also making (the code generated from) quasis act more like actual routines &mdash; which will "accidentally" make `return` behave sensibly inside of a quasi.
